corehq/apps/data_interfaces/tasks.py

```diff
@@ -234,24 +241,34 @@ def task_generate_ids_and_operate_on_payloads(
     payload_id: Optional[str],
     repeater_id: Optional[str],
     domain: str,
-    action: str = '',
+    action,  # type: Literal['resend', 'cancel', 'requeue']  # 3.8+
+    use_sql: bool,
 ) -> dict:
-    repeat_record_ids = _get_repeat_record_ids(payload_id, repeater_id, domain)
-    return operate_on_payloads(repeat_record_ids, domain, action,
+    repeat_record_ids = _get_repeat_record_ids(payload_id, repeater_id, domain,
+                                               use_sql)
+    return operate_on_payloads(repeat_record_ids, domain, action, use_sql,
                                task=task_generate_ids_and_operate_on_payloads)
 
 
 def _get_repeat_record_ids(
     payload_id: Optional[str],
     repeater_id: Optional[str],
     domain: str,
+    use_sql: bool,
 ) -> List[str]:
     if not payload_id and not repeater_id:
         return []
     if payload_id:
-        results = get_repeat_records_by_payload_id(domain, payload_id)
+        if use_sql:
+            return [r.pk for r in get_sql_repeat_records_by_payload_id(
+                domain, payload_id)]
+        else:
+            return [r['id'] for r in get_couch_repeat_records_by_payload_id(
+                domain, payload_id)]
     else:
-        results = iter_repeat_records_by_repeater(domain, repeater_id)
-    ids = [x['id'] for x in results]
-
-    return ids
+        if use_sql:
+            return [r['id'] for r in SQLRepeatRecord.objects.filter(
+                domain=domain, repeater__couch_id=repeater_id).values('id')]
+        else:
+            return [r['id'] for r in iter_repeat_records_by_repeater(
+                domain, repeater_id)]
```

Comment:
nit: might be a bit easier to read if the list comprehension was separated from query construction.
```suggestion
        if use_sql:
            query = SQLRepeatRecord.objects.filter(
                domain=domain, repeater__couch_id=repeater_id).values('id')
        else:
            query = iter_repeat_records_by_repeater(domain, repeater_id)
        return [r['id'] for r in query]
```

Not sure if you could do the same above (in `if payload_id:` block) with something like `get_sql_repeat_records_by_payload_id(...).values('id')`?

Reply:
:thinking: This looks wrong. Why are the Couch IDs "id" and not "_id"? And it seems that unit tests don't cover the return values. I think this might be a bug. I'll add some tests.

Reply:
923453dae76